### PR TITLE
Window transparency and Experimental Web Platform features

### PR DESCRIPTION
**Pull request by katlyn**
Also adds the ability to sync settings to disk if that's something that plugins want to do (currently useless for them though, as they don't load until discord does). Currently only writes changes, does not watch files for changes.

### CODE REVIEW

**Comment by katlyn** [comment]
something I thought about doing but don't have the skills to do is have this pop up when you exit settings, but that's a minor thing (not about hardware accel, but general idea)
image

File: src/Powercord/plugins/pc-settings/components/GeneralSettings.jsx [hunk 1]
```diff
@@ -96,6 +98,25 @@ module.exports = class GeneralSettings extends React.Component {
           >
             Enable Discord Experiments
           </SwitchItem>
+
+          <SwitchItem
+            note={
+              <span>Makes any windows opened by Discord transparent, useful for themeing.<br/><b
+                style={{ color: 'rgb(240, 71, 71)' }}>WARNING:</b> This will break window snapping on Windows. Hardware acceleration must be turned off on linux. Requires restart.</span>
```
Comment (aetheryx) [inline src/Powercord/plugins/pc-settings/components/GeneralSettings.jsx]: (nit) capitalize Linux

File: src/Powercord/settingsManager.js [hunk 2]
```diff
@@ -57,8 +65,17 @@ module.exports = class SettingsManager {
     this._save();
   }
 
-  _save () {
+  async _save () {
     localStorage.setItem(this.category, JSON.stringify(this.config));
+
+    if (this.writeToDisk) {
+      let settingsPath = join(__dirname, '..', '..', 'settings')
```
Comment (aetheryx) [inline src/Powercord/settingsManager.js]: prefer-const

File: src/Powercord/settingsManager.js [hunk 3]
```diff
@@ -57,8 +65,17 @@ module.exports = class SettingsManager {
...
+      if (!(await exists(settingsPath))) {
+        await mkdir(settingsPath)
```
Comment (aetheryx) [inline src/Powercord/settingsManager.js]: missing semis

File: src/patcher.js [hunk 4]
```diff
@@ -6,6 +6,19 @@ const { BrowserWindow, app, session } = electron;
 const electronPath = require.resolve('electron');
 const discordPath = join(dirname(require.main.filename), '..', 'app.asar');
 
+console.log('Patching')
```
Comment (aetheryx) [inline src/patcher.js]: is this intentional? nothing inherently wrong with it, just making sure

File: src/patcher.js [hunk 5]
```diff
@@ -6,6 +6,19 @@ const { BrowserWindow, app, session } = electron;
...
+
+let settings;
+try {
+  settings = require('../settings/general.json')
+  console.log(settings)
+} catch (err) {
+  console.log('catch')
```
Comment (aetheryx) [inline src/patcher.js]: definitely not intentional

File: src/patcher.js [hunk 6]
```diff
@@ -6,6 +6,19 @@ const { BrowserWindow, app, session } = electron;
...
+  settings = {}
+}
+
+const { transparentWindow, experimentalWebPlatform } = settings
```
Comment (aetheryx) [inline src/patcher.js]: missing semi

File: src/patcher.js [hunk 7]
```diff
@@ -15,6 +28,19 @@ class PatchedBrowserWindow extends BrowserWindow {
       opts.webPreferences.nodeIntegration = true;
     }
 
+    if (transparentWindow) {
+      console.log('transparent')
+      opts.transparent = true
+      opts.backgroundColor = '#00000000'
+      opts.frame = false
```
Comment (aetheryx) [inline src/patcher.js]: missing semis + console.log that probably shouldn't be there

File: src/patcher.js [hunk 8]
```diff
@@ -15,6 +28,19 @@ class PatchedBrowserWindow extends BrowserWindow {
...
+    }
+
+    if (experimentalWebPlatform) {
+      if (!opts.webPreferences) opts.webPreferences = {}
+      opts.webPreferences.experimentalFeatures = true
+    }
```
Comment (aetheryx) [inline src/patcher.js:36]: this whole if block (and the one for transparentWindow) can just be moved to the existing if-statement, i.e.  

```js
if (opts.webPreferences && opts.webPreferences.preload) {
  global.originalPreload = opts.webPreferences.preload;
  opts.webPreferences.preload = join(__dirname, 'preload.js');
  opts.webPreferences.nodeIntegration = true;

  if (transparentWindow) {
    opts.transparent = true;
    opts.frame = false;
    opts.backgroundColor = '#00000000';
  }

  if (experimentalWebPlatform) {
    opts.webPreferences.experimentalFeatures = true;
  }
}
```

File: src/patcher.js [hunk 9]
```diff
@@ -15,6 +28,19 @@ class PatchedBrowserWindow extends BrowserWindow {
...
+
+    console.log(opts)
```
Comment (aetheryx) [inline src/patcher.js]: zzz

**Review by aetheryx — CHANGES_REQUESTED** [review]

File: src/patcher.js (hunk 4 above)
Comment (katlyn) [reply src/patcher.js]: Non-intentional, removed with everything else

**Review by aetheryx — APPROVED** [review]

**Pull request merged**